Implement MongoGridFS

RageZBla opened the pull request:
Ticket #4

RageZBla commented on lib/Mongo/MongoGridFSFile.php:
```diff
@@ -71,5 +109,34 @@ public function getBytes() {}
+    private function getChunks()
+    {
+        return $chunks = $this->gridfs->chunks->find(
+            ['file_id' => new \MongoDB\BSON\ObjectID((string) $this->file['_id'])],
```
@alcaeus I don't think I should have to convert type here, I think we got some new types leaking somewhere.

alcaeus replied:
That could very well be. I'll take a look at that once I've merged the other PRs.

RageZBla commented on lib/Mongo/MongoCursor.php:
```diff
@@ -154,6 +155,7 @@ protected function doQuery()
     {
         $options = $this->getOptions() + $this->options;
 
+        //$this->cursor = $this->collection->find(TypeConverter::convertLegacyArrayToObject($this->query), $options);
```
@alcaeus I tried to apply the conversion here but it breaks a lot of tests :fearful:

alcaeus replied:
I thought I had conversion implemented in MongoCursor, turns out I haven't.

alcaeus commented:
Can you rebase this against the current master? Also, I'm closing #14, can you implement all your GridFS work in this PR? That would make merging a lot easier. Thanks!

RageZBla commented:
@alcaeus hey there, I have added the cursor implementation and fixed some issues in file/gridfs. 

When you have few minutes can you give us your feedback.

Thanks!

alcaeus commented on lib/Mongo/MongoGridFS.php:
```diff
@@ -48,14 +56,34 @@ class MongoGridFS extends MongoCollection {
+        }
+
+        $this->database = $db;
+        $this->filesName = $prefix . '.files';
```
You should declare these properties instead of defining them implicitely (and thus public)

alcaeus commented:
Will do - I'd like to finish up type conversion in `MongoCollection` before merging this to ensure the type conversion issues are fixed by then.

alcaeus commented on lib/Mongo/MongoGridFS.php:
```diff
@@ -137,6 +245,95 @@ public function __get($id) {}
+    private function calculateMD5($file, $length)
+    {
+        // XXX: this could be really a bad idea with big files...
+        $data = fread($file, $length);
```
The file pointer may not be at the beginning, so you should seek to position 0 before reading from it. Furthermore, fread only reads up to `$length` bytes from the file. Since you don't know what's behind the stream you shouldn't count on fread returning all data at once.

alcaeus commented on lib/Mongo/MongoGridFS.php:
```diff
@@ -137,6 +245,95 @@ public function __get($id) {}
+        $offset = 0;
+        $i = 0;
+
+        while ($offset < $length) {
```
Instead of looping manually you could use [str_split](https://secure.php.net/manual/en/function.str-split.php) to split up the bytestream into chunks of equal size and inserting them.

RageZBla commented:
@alcaeus I have changed the code taking into account your kind feedback. It is rebased as well.

alcaeus commented:
@RageZBla I took your changes and moved them into #22, I'll add handling for exceptions while inserting chunks later. If you have additional feedback, let me know!

RageZBla commented:
@alcaeus Thanks :+1:
